server/src/main/java/org/folio/shared/index/api/SharedIndexService.java

```diff
@@ -0,0 +1,180 @@
+package org.folio.shared.index.api;
+
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.openapi.RouterBuilder;
+import io.vertx.ext.web.validation.RequestParameter;
+import io.vertx.ext.web.validation.RequestParameters;
+import io.vertx.ext.web.validation.ValidationHandler;
+import java.util.function.Function;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.folio.okapi.common.HttpResponse;
+import org.folio.shared.index.matchkey.MatchKeyMethod;
+import org.folio.shared.index.storage.Storage;
+import org.folio.tlib.RouterCreator;
+import org.folio.tlib.TenantInitHooks;
+import org.folio.tlib.postgres.PgCqlField;
+import org.folio.tlib.postgres.PgCqlQuery;
+
+public class SharedIndexService implements RouterCreator, TenantInitHooks {
+
+  private static final Logger log = LogManager.getLogger(SharedIndexService.class);
+  final Vertx vertx;
+
+  public SharedIndexService(Vertx vertx) {
+    this.vertx = vertx;
+  }
+
+  static String stringOrNull(RequestParameter requestParameter) {
+    return requestParameter == null ? null : requestParameter.getString();
+  }
+
+  Future<Void> putSharedRecords(RoutingContext ctx) {
+    Storage storage = new Storage(ctx);
+    return storage.upsertSharedRecords(ctx.getBodyAsJson()).onSuccess(res -> {
+      JsonArray ar = new JsonArray();
+      // global ids and match keys here ...
+      HttpResponse.responseJson(ctx, 200).end(ar.encode());
+    });
+  }
+
+  Future<Void> getSharedRecords(RoutingContext ctx) {
+    PgCqlQuery pgCqlQuery = PgCqlQuery.query();
```

Comment:
Are we sure we want to use CQL for this rather than straight HTTP parameters?

Reply:
It makes my life a lot more easy.. As 1. It's more powerful we can do AND/OR/NOT and we don't have to agument the OpenAPI spec for every parameter. Note: CQL fields are validated, so anything not known is rejected.

Reply:
Ok, makes sense. We should update the design doc to match this. Also, in the DD we had a /records/{globalId} endpoint that alwaya returned a single rec